Add kubeadm reference docs

munnerz opened the pull request:
Adds the kubeadm advanced reference docs.

Not sure if this is the best location/filename?

@kubernetes/sig-cluster-lifecycle @luxas @errordeveloper

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubernetes/kubernetes.github.io/1312)
<!-- Reviewable:end -->

errordeveloper commented:
@erictune feel free to assign to me

errordeveloper commented:
cc @lukemarsden

lukemarsden commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,95 @@
+- `--cloud-provider`
+
+Currently, `kubeadm` does not provide autodetection of cloud provider.
+This means that networking and persistant volumes are not supported out
```
It's not true that networking isn't supported without a cloud provider, see the kubeadm getting started guide. Also typo "persistant" should be "persistent"

lukemarsden commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,95 @@
+of the box. You can specify a cloud provider using `--cloud-provider`.
+Valid values are the ones support by `controller-manager`, namely `"aws"`,
+`"azure"`, `"cloudstack"`, `"gce"`, `"mesos"`, `"openstack"`, `"ovirt"`,
+`"rackspace"`, `"vsphere"`.
```
We need to document what you need to do to make a cloud providers integration work. @errordeveloper?

errordeveloper replied:
Yes, it's not quite documented right now, and it's not very simple. Last time I've tried in GCE, I failed to get it right.

munnerz replied:
I've not configured the cloud provider using kubeadm yet - can you elaborate on some of the technicalities?

lukemarsden commented:
@munnerz I think this needs to be against `kubernetes:release-1.4` branch

lukemarsden commented:
We should also document what kubeadm is actually doing in this doc. See https://twitter.com/sirhopcount/status/780658942414352384

errordeveloper commented:
> We should also document what kubeadm is actually doing in this doc.

Not sure. Such details can change any time and will get out of date too quickly. I would rather implement `--dry-run`.

lukemarsden commented:
Yes we need to keep the docs up-to-date. But a simple explanation of the architecture of what's going on would be very helpful. I'm happy to write it.

lukemarsden commented:
sorry my mistake this PR needs to be against master

devin-donnelly commented:
Correct. Release-1.4 is now merged, and creating further PRs against it creates a pretty big maintenance burden with respect to keeping release-1.4 up to date and fixing merge conflicts.

Please retarget this PR to master, and sorry for the inconvenience. :(

errordeveloper commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,95 @@
+
+- `--use-kubernetes-version` (default 'v1.4.0') the kubernetes version to initialise
+
+`kubeadm` supports a minimum Kubernetes version of v1.4.0
```
Something like this could be better:
```
`kubeadm` was originally built for Kubernetes version **v1.4.0**, older versions are not supported.
With this flag you can try any future version, e.g.  **v1.5.0-beta.1** whenever it comes out (check [releases page](https://github.com/kubernetes/kubernetes/releases)).
```

luxas commented:
@munnerz Do you have time to get back to this soon?

munnerz commented:
@luxas yep sorry, been a busy couple of days - I'm going through comments now 👍

luxas commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,95 @@
+
+- `--cloud-provider`
+
+Currently, `kubeadm` does not provide autodetection of cloud provider.
```
Here we should document that if the user wants to set cloud-specific settings, populate `/etc/kubernetes/cloud-config.json` and kubeadm will pick it up

luxas commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,95 @@
+This should be a minimum of a /16 so that kubeadm is able to assign /24 subnets
+to each node in the cluster.
+
+- `--service-cidr` (default '100.64.0.0/12')
```
Now `10.12.0.0/12`

munnerz commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,95 @@
+
+- `--external-etcd-cafile` etcd certificate authority certificate file
+- `--external-etcd-certfile` etcd client certificate file
+- `--external-etcd-endpoints` comma separated list of etcd endpoints to use
```
To confirm, does the `Fields that support multiple values will do so either with comma
separation, or by specifying the flag multiple times.` rule apply to `--external-etcd-endpoints`? If not I think it should do for a consistent UX

munnerz commented:
@luxas @errordeveloper @lukemarsden I've gone through and addressed your feedback. Let me know how it reads now! Thanks for all the feedback & sorry for the delay!

luxas commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,126 @@
+- `--service-dns-domain` (default 'cluster.local')
+
+By default, `kubeadm init` will deploy a cluster that assigns services with DNS names
+`<service_name>.<namespace>.cluster.local`. You can use the `--service-dns-domain`
```
it's `<service_name>.<namespace>.svc.cluster.local`

luxas commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,126 @@
+This should be a minimum of a /16 so that kubeadm is able to assign /24 subnets
+to each node in the cluster.
+
+- `--service-cidr` (default '10.12.0.0/12')
```
If you change this, the options have to be reflected in `/etc/systemd/system/kubelet.service.d/10-kubeadm.conf` for the DNS

luxas commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,126 @@
+assign pods IP addresses. By default, the chosen subnet is reserved for carrier
+grade NAT and is highly unlikely to cause a collision.
+
+- `--service-dns-domain` (default 'cluster.local')
```
If you change this, the options have to be reflected in /etc/systemd/system/kubelet.service.d/10-kubeadm.conf for the DNS

luxas commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,126 @@
+- `--service-cidr` (default '10.12.0.0/12')
+
+You can use the `--service-cidr` flag to override the subnet Kubernetes uses to
+assign pods IP addresses. By default, the chosen subnet is reserved for carrier
```
It's not carrier-grade NAT anymore

luxas commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,126 @@
+`"rackspace"`, `"vsphere"`. In order to provide additional configuration for
+the cloud provider, you should create a `/etc/kubernetes/cloud-config.json`
+file manually, before running `kubeadm init`. `kubeadm` will then automatically
+pick those settings up and ensure other nodes are configured correctly.
```
Say that if the user specifies this, he also has to make the corresponding changes in `/etc/systemd/system/kubelet.service.d/10-kubeadm.conf` for it to work
(`--cloud-provider` and `--cloud-config` should be set in that case)

munnerz commented:
@luxas once again, addressed your comments :)

munnerz commented:
Not sure why but the travis build was cancelled midway through!

luxas commented:
You can probably do `git reset --soft HEAD~; git commit -m "message"; git push origin -f kubeadm-reference` in order to run it again.

lukemarsden commented:
let's add a troubleshooting section to include https://kubernetes.slack.com/archives/sig-cluster-lifecycle/p1475438601005828

@errordeveloper on all nodes I added the following
```# cat /etc/sysctl.d/k8s.conf 
net.bridge.bridge-nf-call-ip6tables = 1
net.bridge.bridge-nf-call-iptables = 1
```

[2:48]  
my networking foo is not good enough to know *why* this fixed my issues, just that it did :slightly_smiling_face: so I can't judge whether this should be set by `kubeadm` or that its something that should be documented as a caveat

---
On advice of a member in kubernetes-users, I checked the history (edited)

[9:03]  
It seems net.bridge.bridge-nf-call-iptables was set to 0, causing it to bypass iptables.

[9:03]  
So anyone else experiencing this might want to ensure net.bridge.bridge-nf-call-iptables = 1

lukemarsden commented:
this doc appears to have been merged but not linked to from anywhere? @munnerz?

lukemarsden commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,132 @@
+these to generate certificates for the API server. This uses the IP address
+of the default network interface. If you would like to access the API server
+through a different IP address, or through a hostname, you can override these
+defaults with `--api-advertise-addresses` and `--api-external-dns-names`.
```
OK, but what does each flag actually _do_? Examples would help.

munnerz commented:
@lukemarsden updated, if you could take a look :) I'll squash the commits again once you've read over changes

devin-donnelly commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,144 @@
+# kubeadm reference
+
+This document specifies the advanced options of kubeadm.
```
Doc nit: "This document provides information on how to use kubeadm's advanced options."

("Specifying" is something the user does when they set a parameter, for example)

devin-donnelly commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,144 @@
+
+This document specifies the advanced options of kubeadm.
+
+When run, `kubeadm init` will bootstrap a Kubernetes cluster. This is made up
```
"Running `kubeadm init` bootstraps a Kubernetes cluster. This consists of the following steps:"

devin-donnelly commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,144 @@
+When run, `kubeadm init` will bootstrap a Kubernetes cluster. This is made up
+of a few steps, namely:
+
+1. Generating a token that additional nodes can use to register themselves with
```
"kubeadm generates a token..."

devin-donnelly commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,144 @@
+1. Generating a token that additional nodes can use to register themselves with
+the master in future.
+
+1. Generating a self-signed CA using openssl in order to provision identities
```
"Generating" -> "kubeadm generates"
"in order to" -> "to"

devin-donnelly commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,144 @@
+for each node in the cluster, and for the API server to secure communication
+with clients.
+
+1. Outputting a kubeconfig file for the kubelet to use to connect to the API server,
```
"Outputting" -> "kubeadm outputs"

devin-donnelly commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,144 @@
+1. Outputting a kubeconfig file for the kubelet to use to connect to the API server,
+as well as an additional kubeconfig file for administration.
+
+1. Generating Kubernetes resource manifests for the API server, controller manager
```
"Generating" -> "kubeadm generates"

devin-donnelly commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,144 @@
+components of Kubernetes, and once they are up and running we can use `kubectl`
+to set up/manage any additional components.
+
+1. Any addon components (eg. DNS, discovery) are installed in the aformentioned manner,
```
"kubeadm installs any add-on components, such as DNS or discovery, via the API server."

devin-donnelly commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,144 @@
+
+## Usage
+
+Fields that support multiple values will do so either with comma
```
Avoid future tense ("will do so").
"Fields that support multiple values do so by either separating each value with commas, or by specifying the flag multiple times with each value."

devin-donnelly commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,144 @@
+
+- `--token`
+
+By default, `kubeadm init` will automatically generate the token used to initialise
```
Future tense. See comment above. Use present tense.
"By default, `kubeadm init` automatically generates..."

devin-donnelly commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,144 @@
+
+- `--service-dns-domain` (default 'cluster.local')
+
+By default, `kubeadm init` will deploy a cluster that assigns services with DNS names
```
Future tense. See comment above. Use present tense.

devin-donnelly commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,144 @@
+- `--external-etcd-certfile` etcd client certificate file
+- `--external-etcd-keyfile` etcd client key file
+
+By default, `kubeadm` will deploy a single node etcd cluster on the master
```
Future tense. See comment above. Use present tense.

devin-donnelly commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,144 @@
+the cloud provider, you should create a `/etc/kubernetes/cloud-config.json`
+file manually, before running `kubeadm init`. `kubeadm` will then automatically
+pick those settings up and ensure other nodes are configured correctly.
+You will also need to set the `--cloud-provider` and `--cloud-config` parameters
```
"You will also need to" -> "You must also"

luxas commented:
@munnerz Can you update this one and address @devin-donnelly's comments?

munnerz commented:
@devin-donnelly @errordeveloper I've updated the PR with changes!

luxas commented on docs/user-guide/kubeadm-reference.md:
```diff
@@ -0,0 +1,143 @@
+each new node. If you would like to manually specify this token, you can use the
+`--token` flag. The token must be of the format '<6 character string>.<16 character string>'.
+
+- `--use-kubernetes-version` (default 'v1.4.0') the kubernetes version to initialise
```
default will now be v1.4.1

luxas commented:
Many thanks! Possible to squash now and rebase to current master? Until then I can't merge...

This LGTM, if anyone wants more changes to this now, I think that should be done in a separate PR

munnerz commented:
@luxas thanks for reviewing - I've updated the default version, squashed and rebased!
